src/core/ActionWithVector.cpp

```diff
@@ -528,131 +273,77 @@ void ActionWithVector::getNumberOfTasks( unsigned& ntasks ) {
-}
-
-void ActionWithVector::getNumberOfStreamedQuantities( const std::string& headstr, unsigned& nquants, unsigned& nmat, unsigned& maxcol, unsigned& nbookeeping ) {
-  setupStreamedComponents( headstr, nquants, nmat, maxcol, nbookeeping );
-  if( action_to_do_after ) action_to_do_after->getNumberOfStreamedQuantities( headstr, nquants, nmat, maxcol, nbookeeping );
-}
-
-void ActionWithVector::getSizeOfBuffer( const unsigned& nactive_tasks, unsigned& bufsize ) {
-  for(int i=0; i<getNumberOfComponents(); ++i) { getPntrToComponent(i)->bufstart=bufsize; bufsize += getPntrToComponent(i)->data.size(); }
-  if( action_to_do_after ) action_to_do_after->getSizeOfBuffer( nactive_tasks, bufsize );
-}
-
-void ActionWithVector::getNumberOfStreamedDerivatives( unsigned& nderivatives, Value* stopat ) {
-  std::string c=getFirstActionInChain()->getLabel();
-  for(unsigned i=0; i<getNumberOfArguments(); ++i) {
-    if( !getPntrToArgument(i)->ignoreStoredValue(c) ) {
-      if( getPntrToArgument(i)==stopat ) return;
-      nderivatives += getPntrToArgument(i)->getNumberOfValues();
-    }
-  }
-  if( getNumberOfAtoms()>0 ) nderivatives += 3*getNumberOfAtoms() + 9;
-  // Don't do the whole chain if we have been told to stop early
-  if( stopat && stopat->getPntrToAction()==this ) return;
-
-  if( action_to_do_after ) action_to_do_after->getNumberOfStreamedDerivatives( nderivatives, stopat );
-}
-
-bool ActionWithVector::getNumberOfStoredValues( Value* startat, unsigned& nvals, const unsigned& astart, const std::vector<Value*>& stopat ) {
-  for(unsigned j=astart; j<stopat.size(); ++j) {
-    if( stopat[j] && (stopat[j]->getPntrToAction()==this || (stopat[j]->getPntrToAction())->checkForDependency(this)) ) return true;
-  }
-
-  std::string c=getFirstActionInChain()->getLabel();
-  for(unsigned i=0; i<getNumberOfArguments(); ++i) {
-    if( !getPntrToArgument(i)->ignoreStoredValue(c) ) {
-      for(unsigned j=astart; j<stopat.size(); ++j) {
-        if( getPntrToArgument(i)==stopat[j] ) return true;
-      }
-      nvals += getPntrToArgument(i)->getNumberOfValues();
-    }
-  }
-  if( startat->getPntrToAction()!=this && getNumberOfAtoms()>0 ) return false;
-
-  if( action_to_do_after ) return action_to_do_after->getNumberOfStoredValues( startat, nvals, astart, stopat );
-  return false;
 }
 
 void ActionWithVector::runTask( const unsigned& current, MultiValue& myvals ) const {
-  if( isActive() ) {
-    myvals.setTaskIndex(current); myvals.vector_call=true; performTask( current, myvals );
+  const ActionWithMatrix* am = dynamic_cast<const ActionWithMatrix*>(this);
+  myvals.setTaskIndex(current); myvals.vector_call=true; performTask( current, myvals );
+  for(unsigned i=0; i<getNumberOfComponents(); ++i) {
+    const Value* myval = getConstPntrToComponent(i);
+    if( am || myval->hasDerivatives() ) continue;
+    Value* myv = const_cast<Value*>( myval );
+    if( getName()=="RMSD_VECTOR" && myv->getRank()==2 ) continue;
+    myv->set( current, myvals.get( i ) );
```

Comment:
This is a const function in which non-const things happen, it is called from non-const functions(runAllTasks and checkForForces), so it does not need the const modifier.
It may be easier to unflag this from its constness and change to code to something like, so there is not need to `const_cast` the values to modify them:
```suggestion
  myvals.setTaskIndex(current);
  myvals.vector_call=true;
  performTask( current, myvals );
  if (const ActionWithMatrix* am = dynamic_cast<const ActionWithMatrix*>(this);!am){
    const bool IamRMSD_VECTOR=getName()=="RMSD_VECTOR";
    for(unsigned i=0; i<getNumberOfComponents(); ++i) {
      if(Value* myval = getPntrToComponent(i);
          !myval->hasDerivatives() && 
          !( IamRMSD_VECTOR && myval->getRank()==2 )) {
        myval->set( current, myvals.get( i ) );
      }
     }
```

I also moved some conditions outside the loop and changed the scope of `am`

Maybe the discrimination for `RMSD_VECTOR` can be done in it? Or with an ad-hoc overloaded method `virtual bool skipIf(Value *)const{return true};`?

Then for the parallelization, we can manipulate MultiValue into a Data sitter for cpu/gpu adresses